File: flink-monitor/src/main/scala/ch/ethz/infsec/trace/ParsingFunction.scala
package ch.ethz.infsec.trace

import java.util

import ch.ethz.infsec.monitor.Fact
import ch.ethz.infsec.trace.parser.TraceParser
import org.apache.flink.api.common.functions.FlatMapFunction
import org.apache.flink.streaming.api.checkpoint.ListCheckpointed
import org.apache.flink.util.Collector

class ParsingFunction(var parser: TraceParser) extends FlatMapFunction[String, Fact] with ListCheckpointed[TraceParser] {
  override def flatMap(line: String, collector: Collector[Fact]): Unit = parser.parseLine(collector.collect(_), line)

  override def snapshotState(checkpointId: Long, timestamp: Long): util.List[TraceParser] =
    util.Collections.singletonList(parser)

  override def restoreState(state: util.List[TraceParser]): Unit = {
    assert(state.size() == 1)
    parser = state.get(0)
  }
}

File: flink-implementation/src/main/java/ch/ethz/infsec/util/ParsingFunction.java
package ch.ethz.infsec.util;


import java.util.*;

import ch.ethz.infsec.monitor.Fact;
import ch.ethz.infsec.trace.parser.TraceParser;
import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.streaming.api.checkpoint.ListCheckpointed;
import org.apache.flink.util.Collector;

public class ParsingFunction implements FlatMapFunction<String, Fact>, ListCheckpointed<TraceParser> {
        TraceParser parser;
    public ParsingFunction(TraceParser tp){
        this.parser = tp;
    }

    @Override
    public void flatMap(String line, Collector<Fact> collector) throws Exception {
        this.parser.parseLine(collector::collect, line);
    }

    @Override
    public List<TraceParser> snapshotState(long checkpointId, long timestamp) throws Exception {

        return new ArrayList<>(Collections.singleton(parser));
    }

    @Override
    public void restoreState(List<TraceParser> state) throws Exception {
        assert(state.size() == 1);
        parser = state.get(0);
    }
}
